### ISSUE TITLE
Best Practices for Importing Private Keys from Desktop GPG clients.

### ISSUE THREAD

**Issue by zmanian** [open-keychain/open-keychain#975]
I've run into a challenge of what to recommend to new users as methodology of importing an existing private key from desktops to their mobile for use on OpenKeychain. 

We can all agree that sending a private key over a cloud service is a poor operational practice.

My current practice is to use a subkey on my mobile device and my intention to rotate those subkeys once per year. I transferred that subkey over ssh to my mobile device. My eventual goal is to move to a Yubikey type device to managing my mobile subkey.

Proposed recommendation to new users:

- Export your private key to a fille.
 Mailvelope: Export all keys -> keys.asc 
 gpg: gpg --export-secret-key ####### --armor > keys.asc

- Connect your android device via USB
 Transfers keys.asc to Android device

- Install a file navigator. 
Perhaps using https://play.google.com/store/apps/details?id=org.openintents.filemanager&hl=en

- Import key into OpenKeychain

**Comment by ClashTheBunny** [comment]
If I put armor after the command, it doesn't ASCII armor the file, it goes before the export option.

**Comment by Valodim** [comment]
I added a FAQ with some points to the [website](http://www.openkeychain.org/help/), that should help. This was just from the top of my head, hope I didn't mess anything up.

**Comment by timbray** [comment]
​That’s good!  I would make slight adjustments:

1. The point that the key is passphrase-protected is crucially important
and I think should be right at the top and the first paragraph. e.g.: “Your
private key is passphrase-protected and an attacker will gain no useful
information from your exported key file unless they can guess it.”  Then
immediately say that it’s very important that the passphrase has never
traveled over the network.
2. A high-quality 20-char-plus password is really onerous to use.  There
are two acceptable strategies to avoid this:
- use a password manager to make & remember the password for you.  Use one
that does not have a web interface, so the password manager password
doesn't cross the wire (e.g. 1password)
- use a Yubikey, which you already explain lower down on the page.

On Mon, Feb 16, 2015 at 6:41 AM, Vincent <notifications@github.com> wrote:

> I added a FAQ with some points to the website
> <http://www.openkeychain.org/help/>, that should help. This was just from
> the top of my head, hope I didn't mess anything up.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/open-keychain/open-keychain/issues/975#issuecomment-74517609>
> .
>



-- 
- Tim Bray (If you’d like to send me a private message, see
https://keybase.io/timbray)

**Comment by Valodim** [comment]
1. I agree, will change when I have time for it
2. The whole point of those instructions is that this particular passphrase is used only once for the transfer. I take note that this is mentioned too far down, and change the text accordingly. :+1:

**Comment by Valodim** [comment]
Ok I changed it a bit, highlighted two critical parts. I also added some more exposition about the meaning of "if the passphrase is as difficult to guess as the key", and give an example of how a passphrase should look. Can someone have a look if I hit a good ratio between understandable and reasonably accurate?

**Comment by timbray** [comment]
Detail points:
- 1st comma after “isn’t” is superfluous
- Slightly awkward sentence in #3.  Refactoring a bit: “OpenKeychain’s key
database can’t be accessed by other apps unless they have root privileges
(which means you’ve rooted your device and and presumably know what you’re
doing.)”
- What proportion of mobiles have SD cards any more?  I haven’t had one in
years, but maybe my experience is weird. I just surveyed my family and none
of the 3 phones have one. So I worry a little bit about leading with that
as the preferred option.
- Never heard of PushBullet. Maybe mention SpiderOak?

BTW, I regard the 16-char random passwords generated by 1Password as
sufficiently strong for all practical purposes, as in “hard enough to
brute-force that it’s uneconomic for the attacker to try”.  Maybe you don’t
agree specifically, but a recommendation to use a password generator rather
than trying to invent one yourself might be worth making.

On Tue, Feb 17, 2015 at 4:26 AM, Vincent <notifications@github.com> wrote:

> Ok I changed it a bit, highlighted two critical parts. I also added some
> more exposition about the meaning of "if the passphrase is as difficult to
> guess as the key", and give an example of how a passphrase should look. Can
> someone have a look if I hit a good ratio between understandable and
> reasonably accurate?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/open-keychain/open-keychain/issues/975#issuecomment-74660209>
> .
>



-- 
- Tim Bray (If you’d like to send me a private message, see
https://keybase.io/timbray)

**Comment by Valodim** [comment]
- aight
- aight
- it was the first safe option that came to mind, but you're right sd card readers aren't exactly common anymore. I have no strong feelings about this, opinions anyone?
- never heard of SpiderOak, so uh... third opinion? doesn't seem important either way

Your last point shows an important thing, which I worry hasn't been clear enough: "it's hard enough to be impractical" is specifically *not* what we are aiming for here.

The transfer of a secret key is a typical "weakest link" situation for your secret key. If you choose a passphrase which has any less bits of entropy than your secret key, you effectively make this key the difference in bits easier for an attacker to crack. Going by your example of a 16-char random password generated by 1Password, you have something like 85 bits of entropy, which means your 4096 bit RSA key has just been demoted to a ~1200 bits RSA key.

**Comment by denydias** [comment]
Hi,

It would be way more secure to export only the private subkeys (encrypt and an additional sign key) in the key origin system. So the private part of the master key will never leave the origin system. Unfortunatly, OpenKeychain does not support this as it refuses to edit keys with striped master key. Without that feature a user has no way to, for instance, change the sign subkey passphrase to a more convenient one under tiny touch keyboards.

**Comment by dschuermann** [comment]
@denydias We want to be usable. Using subkeys isn't usable because you can't certify other keys without having the master key on your smartphone. We don't want to encourage average users to go this way.
This said, please open a new issue regarding the edit of stripped keys, we should fix that.
And please don't mention APG, it's dead.

**Comment by denydias** [comment]
Fine then, @dschuermann.

I'm not going to open an issue for this. I'm experienced enough to workaround the software limitation and use the app the way I want, not the way you think is right and want to lecture me or to forbid my mention to anything.

I just feel sorry for your tone as it may also go towards other users free will.

Sorry for bother you and your project.

**Comment by plsng** [comment]
@denydias OpenKeychain is free software, it's impossible for the developers to go against the users' free will. If you want to see something the original developers don't want to include, you can create your own version of the application and share the changes with others who might also think like you do.

**Comment by Valodim** [comment]
Hi!

so we should address the point "APG is dead" in a little more detail.

For brevity, I will only address the most important one: APG is insecure. In a lot of places, it implements features without proper regard for security. For instance, everyone can add an uncertified user id or even subkey to any key and it will show up in APG like any other, because there simply *are no checks* for it. This has lead to cases of people complaining that "I can't import my key in OpenKeychain, this worked just fine in APG why can't you be more like APG!"

I spent **a lot** of time writing several thousand lines of code in OpenKeychain which make sure all keys are valid, rejecting bad keys or parts of keys. Now I don't blame those people for not knowing implementation details, and I am not saying we don't have bugs. But over time, this has left me and dominik a little on edge about people who complain about security aspects, pointing at APG for "doing it right".

Still, we shouldn't go around saying "APG is dead" without context like that. Sorry for that.

**Comment by denydias** [comment]
Hi, @valodim.

Now this is a answer. I'm deeply grateful to you for taking the time to contextualize the issue. I wish @dschuermann could learn something with you on how to treat users.

I'm in 100% agreement with your explanation. I also agree with @dschuermann that OpenKeychain must be usable. The point is that any software has the role to give users the ability to be used to the extent she/him want it to be, not the opposite.

Let's take out of the context OpenKeychain and APG for a moment and stick just with GPG itself. The original, pure C code that implements OpenPGP in a myriad of platforms. A security wise user can have it's own methods to deal with private keys and subkeys. A good (not the only), security wise approach is to avoid carrying on a master private key everywhere, keeping it always and only in an offline system.

In the event of a 'sign party', this user could get a piece of paper from each other user he wants to sign. This paper has the other user key fingerprint and the owner signature made with a pen. Then, the user who is going to sign the keys from others come back to home, sit in front the airgapped system, loads the private key, sign the third party keys, export them, go back to a online system and publish the signed keys.

This looks like a lot of work, but it's quite usable and works like a charm. GPG handles that procedure just fine. It has the features to implement the user policy to sign third party keys in a offline system.

In the other hand, we could have a non security wise user who don't even bother to add a passphrase to her/his own master key and copy this key to multiple, online systems. GPG also handles that just fine.

Is GPG insecure? Not at all. The user choices are.

This is a typical case where not having to transfer the master key to a mobile device comes in place. OpenKeychain don't need to force its users to carry their master keys around if they don't want to. Of course this user is going to loose the ability to certify other keys within mobile devices, but this is not OpenKeychain's fault. It's a policy the user choose to implement and she/he is willing to deal with that.

I also agree with you that OpenKeychain MUST 'make sure all keys are valid, rejecting bad keys or parts of keys.' This is good and expected for any good GPG implementation. But among these concerns, the ability to change a password for a striped down master key must be preserved.

I see that @dschuermann has filled #1085, which I refused to open as I got very pissed off with his aggressive tone to my concern. I thank him for having filled that.

**Comment by dschuermann** [comment]
That's why I should avoid getting into these kind of discussions, they take up precious development time (this is still my spare time, no money involved here). Not sure why you are pissed, in comparison to other crypto implementations nowadays, we still allow many advanced scenarios.

I don't agree with your most prominent statement: "Is GPG insecure? Not at all. The user choices are."

GnuPG together with it's user base is the problem why the OpenPGP standard has not reached a critical mass. Exactly this kind of over-paranoid handling of secret keys scares average users. They don't understand one single word you are saying and thus will never even bother signing other keys.
The only reason this paranoid handling of master keys exists is because of the Web-of-Trust model. As you are obviously a defender of this model, may I ask you how many trust signature (http://tools.ietf.org/html/rfc4880#section-5.2.3.13 ) have you created? And/Or how many keys have a trust level associated in you GnuPG installation?
If your answer is 0 or a small number, the web of trust also failed for you.
Have you ever contacted someone that you validated via a transitive trust relationship?

**Comment by denydias** [comment]
Gosh, @dschuermann! You're such a jerk, just like me!

I should not bother to reply to you, but I'll bite. So here I go.

> That's why I should avoid getting into these kind of discussions, they take up precious development time (this is still my spare time, no money involved here).

Then you're obviously in the wrong place. This issue has one label only: a yellow one with 'discussion' written into it.

As for the matter, this also takes precious users time (this is still my spare time, no money involved here).

> Not sure why you are pissed, in comparison to other crypto implementations nowadays, we still allow many advanced scenarios.

I wasn't pissed with your pretty valid arguments. I was pissed with the prepotent tone of your reply. This is past now. **Now I'm pissed with you personally.** This is my last interaction with you in our short 'online relationship' life.

Again, I agree with you that OpenKeychain allows many advanced use case scenarios. What you haven't understood yet is that I'm not talking about these implemented scenarios. I was trying to talk about the ones not in place.

> I don't agree with your most prominent statement: "Is GPG insecure? Not at all. The user choices are."

It's hard to decide on what part you do not agreed. Is GPG itself insecure and user choices are secure? Well, I'll not ask you to develop on this, as I don't care with your answer. It's enough to me to know that my statement is backed up by the one rule of life: wrong decisions renders the most well designed system useless. I'm not talking that GPG or its WoT are those well designed systems, far from it. I'm talking that they allow users who wants to implement secure choices.

> GnuPG together with it's user base is the problem why the OpenPGP standard has not reached a critical mass. Exactly this kind of over-paranoid handling of secret keys scares average users. They don't understand one single word you are saying and thus will never even bother signing other keys.

Well, besides the fact just called 'dumbasses' the user base of the system you are developing on top of it, you also just called your users 'dumb' enough so they are unable to learn how the system is supposed to work. Congratulations, smart guy. 

But think about it for a while in your next poop: if they are dumb enough to 'don't understand one single word you (me) are saying', how this is exactly supposed to 'scares average users'. Ignorance is bliss.

Extending even further: if they are dumb enough to 'don't understand one single word you (me) are saying and thus will never even bother signing other keys', why there is a sign feature in OpenKeychain in the first place? Just drop it off and it's fine.

> The only reason this paranoid handling of master keys exists is because of the Web-of-Trust model. As you are obviously a defender of this model..

The only reason this handling of master keys exists is because a user wants it that way. The WoT is another thing, out of the individual scope. In OpenSSL world, how do you thrust a Root CA that keeps its master CA certs in an online fashion? It's supposed to be airgapped, although some of them may brake this fundamental security policy. If it's good enough and desired to a business to keep a master private key or root CA offline, why it's not good enough to an individual to do the very same? Again, I don't care about your written answer.

> may I ask you how many trust signature (http://tools.ietf.org/html/rfc4880#section-5.2.3.13 ) have you created? And/Or how many keys have a trust level associated in you GnuPG installation?
If your answer is 0 or a small number, the web of trust also failed for you.

Sure! I only sign signatures from the coworkers I work directly with and revoke the signatures of the ones that leave. At this moment they are three. And yes, WoT failed to me and to [many](https://lists.torproject.org/pipermail/tor-talk/2013-September/030235.html) out [there](http://secushare.org/PGP).

> Have you ever contacted someone that you validated via a (potentially transitive) trust relationship?

Yeap. I do this every single work day of my life.

**Comment by dschuermann** [comment]
Just for others to understand:
* @denydias is not using the web of trust, there are no trust relationships. He only certified keys directly, meaning: he only trusted himself
* Yes, you are kind of a CA in the web of trust model, but only if others are defining a trust relationship to you. As most people I know don't do this, I don't see a reason to have the same hard security requirements as X.509 CAs.
* Yes, I assume my users are dumb in the sense that they don't know most notations or concepts. Nobody explained them in school. The Johnny papers have shown that this assumption is valid (http://scholar.google.de/scholar?hl=de&as_sdt=0,5&q=why+johnny+can't+encrypt)

@denydias
Again: Not sure why pissed _personally_ . This is how security works: Discussing until one finds a loophole. Looks like we also agree on many points :tada:

**Comment by mjlong2014** [comment]
How do i export my private/public key file? When I choose export and try to select a file, it wont let me choose a file to export it to.  I've tried selecting my google drive and my phone, but i select any file to export to.

**Comment by dschuermann** [comment]
@mjlong2014 Which Android version and OpenKeychain version are you using?

**Comment by mjlong2014** [comment]
I figured it out. It exported the file to the openkeychain folder on my
phone and from there i can share it to where ever i want. Thank you for
getting back to me tho!!! Love the app!!!

**Comment by danielhatton** [comment]
A footnote: there's a lot of talk in this thread about a web of trust, but I'm not sure the concept is relevant.  Surely signing a PGP key only indicates "I've seen convincing evidence that the biographical information recorded in this key is correct biographical information of the keyholder", not "I trust the keyholder to be conscientious in verifying other people's keys"?

**Comment by sarnold** [comment]
Given the note on the readme about no more active development, can anyone comment on the state of the "last" release wrt this thread and importing a stripped down gpg key on the phone side?  This still seems to be the GPG app of choice in F-Droid.  Thanks in advance...
